src/asl.ts

```diff
@@ -5261,42 +5270,59 @@ export namespace ASLGraph {
       operator === ">=" ||
       operator === "<="
     ) {
+      // left is a condition and right is any value
       if (ASLGraph.isConditionOutput(left)) {
         return (
           ASLGraph.booleanCompare(left, right, operator) ?? ASL.falseCondition()
         );
       }
+
+      // returns comparisons if there are valid comparisons to return.
+      // comparisons return undefined when the right is a literal and
+      // not the correct type to compare (aString === aNumber) or when the
+      // operator doesn't make sense for the type (ex: less than boolean)
+      // when L and R are paths: typeof L === "string" && stringCompare(L, R)
       const condition = ASL.or(
         ASLGraph.nullCompare(left, right, operator),
         ASLGraph.stringCompare(left, right, operator),
         ASLGraph.booleanCompare(left, right, operator),
         ASLGraph.numberCompare(left, right, operator)
       );
 
-      if (ASLGraph.isJsonPath(right)) {
-        ASL.or(
-          // a === b while a and b are both not defined
-          ASL.not(
-            ASL.and(ASL.isPresent(left.jsonPath), ASL.isPresent(right.jsonPath))
-          ),
-          // a !== undefined && b !== undefined
-          ASL.and(
-            ASL.isPresent(left.jsonPath),
-            ASL.isPresent(right.jsonPath),
-            // && a [op] b
-            condition
-          )
-        );
-      }
-      return ASL.and(ASL.isPresent(left.jsonPath), condition);
+      return ASL.or(
+        // for ==/===
+        // if right is undefined literal, check isNotPresent(left)
+        // if right is a variable, check isNotPresent(left) && isNotPresent(right)
+        ASLGraph.undefinedCompare(left, right, operator),
```

Comment:
What happens if this returns `undefined`?

Reply:
We are trying to collect the comparisons we need, when the values are literals, we only do the comparisons of the type of the literal value. When the values are variable, we must do all of the comparisons to try to find a `true`.

So the `*compare` methods return a Condition when we need a runtime condition and `undefined` when we can assume the comparison is false.

Reply:
I still don't understand what happens when `ASL.or(undefined, ..)`

Reply:
ohh, `or` and `and` filter out undefined. if all undefined are passed to `or`, it returns falsy condition, if all undefined are passed to AND it returns a truthy condition.